onnx/checker.cc

```diff
@@ -0,0 +1,247 @@
+    case TypeProto::kTensorType: {
+      const auto& type = value_info.type().tensor_type();
+      enforce_has_field(type, elem_type);
+      enforce_has_field(type, shape);
+    } break;
+    case TypeProto::kSparseTensorType: {
+      const auto& type = value_info.type().sparse_tensor_type();
+      enforce_has_field(type, elem_type);
+      enforce_has_field(type, shape);
+    } break;
+    default:
+      fail("Unrecognized type value case: ", value_case);
+  }
+}
+
+void check_tensor(const TensorProto& tensor, int ir_version) {
+  enforce_has_field(tensor, data_type);
+  if (tensor.data_type() == TensorProto::UNDEFINED) {
+    fail("setting data_type field to UNDEFINED is not allowed");
+  }
+
+  enforce_has_repeated_field(tensor, dims);
+
+  int num_value_fields = 0;
+
+  const char* value_field = nullptr;
+
+#define check_data_field(field)             \
+  bool has_##field = tensor.field().size(); \
+  if (has_##field) {                        \
+    ++num_value_fields;                     \
+    value_field = #field;                   \
+  }
+
+  check_data_field(float_data);
+  check_data_field(int32_data);
+  check_data_field(string_data);
+  check_data_field(int64_data);
+  check_data_field(raw_data);
+  check_data_field(double_data);
+  check_data_field(uint64_data);
+
+#undef check_data_field
+
+  if (num_value_fields != 1) {
+    fail("TensorProto should contain one and only one value field.");
+  }
+  if (has_raw_data) {
+    if (tensor.data_type() == TensorProto::STRING) {
+      fail("STRING data should not be stored in raw_data field");
+    }
+  } else {
+#define check_field(field)               \
+  if (!has_##field) {                    \
+    fail(                                \
+        "values of data_type '",         \
+        tensor.data_type(),              \
+        "' should be stored in field '", \
+        #field,                          \
+        "' instead of '",                \
+        value_field,                     \
+        "'");                            \
+  }
+
+    switch (tensor.data_type()) {
+      case TensorProto::FLOAT:
+      case TensorProto::COMPLEX64:
+      case TensorProto::COMPLEX128:
+        check_field(float_data);
+        break;
+
+      case TensorProto::DOUBLE:
+        check_field(double_data);
+        break;
+
+      case TensorProto::INT32:
+      case TensorProto::UINT16:
+        check_field(int32_data);
+        break;
+
+      case TensorProto::INT64:
+        check_field(int64_data);
+        break;
+
+      case TensorProto::UINT32:
+      case TensorProto::UINT64:
+        check_field(uint64_data);
+        break;
+
+      case TensorProto::STRING:
+        check_field(string_data);
+        break;
+
+      default:
+        fail("Unrecognized data_type: ", tensor.data_type());
+    }
+  }
+
+#undef check_field
+}
+
+void check_attribute(const AttributeProto& attr, int ir_version) {
+  enforce_non_empty_field(attr, name);
+
+  if (ir_version >= 0x00000002) {
```

Comment:
I think checker should always take the latest version and we should have a converter before that